Cross Agent Dependencies don't respond correctly to unavailable resources

# Problem

The following problem was observed in an environment with many unavailable resources

``` python
2020-11-05 21:03:43,519 ERROR    inmanta.util An exception occurred while handling a future: 1 validation error for Event
change
  none is not an allowed value (type=type_error.none.not_allowed)
Traceback (most recent call last):
  File "/opt/inmanta/lib64/python3.6/site-packages/inmanta/util.py", line 256, in handle_result
    f.result()
  File "/opt/inmanta/lib64/python3.6/site-packages/inmanta/agent/agent.py", line 262, in execute
    for x in self.dependencies
  File "/opt/inmanta/lib64/python3.6/site-packages/inmanta/agent/agent.py", line 262, in <dictcomp>
    for x in self.dependencies
  File "/opt/inmanta/lib64/python3.6/site-packages/pydantic/main.py", line 346, in __init__
    raise validation_error
pydantic.error_wrappers.ValidationError: 1 validation error for Event
change
  none is not an allowed value (type=type_error.none.not_allowed)
```

# Analysis

From the error, it is clear that the 'change' field is None for a 'ResourceAction'.

The only potential location I could find where this may happen is in `RemoteResourceAction` 

https://github.com/inmanta/inmanta/blob/b6a01eb15744f1dafa31bd023eb3061093147e03/src/inmanta/agent/agent.py#L372

If this condition is negative, the value remains None.

# Mypy
Mypy already knew:

It might be interesting to also adres the other potential type problem here

```
src/inmanta/agent/agent.py:260: error: Argument "status" to "Event" has incompatible type "Optional[ResourceState]"; expected "ResourceState"
src/inmanta/agent/agent.py:260: error: Argument "change" to "Event" has incompatible type "Optional[Change]"; expected "Change"
```

# Fix 

I would propose to change the default here https://github.com/inmanta/inmanta/blob/b6a01eb15744f1dafa31bd023eb3061093147e03/src/inmanta/agent/agent.py#L94
to `Change.NoChange`

# Testing

It should be verified if this condition is caused by Unavailable resources.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,6 +16,7 @@
 - Ensure the compiler service takes into account the environment variables set on the system (#2413)
 - Fix `--server_address` option on `inmanta export` (#2514)
 - Handle failure in an event handler consistently for local and non-local agents (#2509)
+- Fix for cross agent dependencies responding to unavailable resources (#2501)
 
 ## Upgrade notes
 - Ensure the database is backed up before executing an upgrade.

--- a/src/inmanta/agent/agent.py
+++ b/src/inmanta/agent/agent.py
@@ -92,7 +92,7 @@ class ResourceAction(object):
         self.running: bool = False
         self.gid: uuid.UUID = gid
         self.status: Optional[const.ResourceState] = None
-        self.change: Optional[const.Change] = None
+        self.change: Optional[const.Change] = const.Change.nochange
         self.undeployable: Optional[const.ResourceState] = None
         self.reason: str = reason
         self.logger: Logger = self.scheduler.logger

--- a/src/inmanta/server/services/resourceservice.py
+++ b/src/inmanta/server/services/resourceservice.py
@@ -546,6 +546,8 @@ class ResourceService(protocol.ServerSlice):
             for agent, resource_id in waiting_agents:
                 aclient = self.agentmanager_service.get_agent_client(env.id, agent)
                 if aclient is not None:
+                    if change is None:
+                        change = const.Change.nochange
                     await aclient.resource_event(env.id, agent, resource_id, send_events, status, change, changes)
 
         return 200

--- a/tests/agent_server/test_send_events.py
+++ b/tests/agent_server/test_send_events.py
@@ -25,7 +25,7 @@ from inmanta import config, const
 from inmanta.agent.agent import Agent
 from inmanta.server import SLICE_AGENT_MANAGER
 from inmanta.util import get_compiler_version
-from utils import _wait_until_deployment_finishes, retry_limited
+from utils import _wait_until_deployment_finishes, log_doesnt_contain, retry_limited
 
 logger = logging.getLogger("inmanta.test.send_events")
 
@@ -463,3 +463,89 @@ async def test_consistenly_handle_failure_in_process_events(
     await assert_deployment_status(res_id=res_id_key2, expected_state=const.ResourceState.skipped)
     # Skipped due to failed local dependency
     await assert_deployment_status(res_id=res_id_key1, expected_state=const.ResourceState.skipped)
+
+
+@pytest.mark.asyncio(timeout=15)
+async def test_send_events_cross_agent_unavailable(
+    resource_container, environment, server, client, clienthelper, no_agent_backoff, async_finalizer, caplog
+):
+    """
+    Having unavailable cross agent dependencies shouldn't result in invalid Event objects (#2501)
+    """
+    agentmanager = server.get_slice(SLICE_AGENT_MANAGER)
+
+    config.Config.set("config", "agent-deploy-interval", "0")
+    config.Config.set("config", "agent-repair-interval", "0")
+
+    resource_container.Provider.reset()
+
+    agent2 = Agent(hostname="node2", environment=environment, agent_map={"agent2": "localhost"}, code_loader=False)
+
+    async_finalizer.add(agent2.stop)
+    await agent2.add_end_point_name("agent2")
+    await agent2.start()
+    await retry_limited(lambda: len(agentmanager.sessions) == 1, 10)
+
+    agent = Agent(hostname="node1", environment=environment, agent_map={"agent1": "localhost"}, code_loader=False)
+    async_finalizer.add(agent.stop)
+    await agent.add_end_point_name("agent1")
+    await agent.start()
+    await retry_limited(lambda: len(agentmanager.sessions) == 2, 10)
+
+    version = await clienthelper.get_version()
+
+    res_id_1 = "test::Resource[agent1,key=key1],v=%d" % version
+    res_id_2 = "test::ResourceNoHandler[agent2,key=key2],v=%d" % version
+    res_id_3 = "test::Resource[agent2,key=key3],v=%d" % version
+    # resource 1 requires an available and an unavailable resource
+    resources = [
+        {
+            "key": "key1",
+            "value": "value1",
+            "id": res_id_1,
+            "send_event": True,
+            "requires": [res_id_2, res_id_3],
+            "purged": False,
+        },
+        {
+            "key": "key2",
+            "value": "value2",
+            "id": res_id_2,
+            "send_event": True,
+            "requires": [],
+            "purged": False,
+        },
+        {
+            "key": "key3",
+            "value": "value3",
+            "id": res_id_3,
+            "send_event": True,
+            "requires": [],
+            "purged": False,
+        },
+    ]
+    await clienthelper.put_version_simple(resources, version)
+
+    # do a deploy
+    result = await client.release_version(environment, version, True, const.AgentTriggerMethod.push_full_deploy)
+    assert result.code == 200
+
+    result = await client.get_version(environment, version)
+    assert result.code == 200
+
+    await _wait_until_deployment_finishes(client, environment, version)
+
+    async def check_resource_state(environment, id, state):
+        result = await client.get_resource(tid=environment, id=id, logs=False, status=True)
+        assert result.code == 200
+        return result.result["status"] == state
+
+    assert await check_resource_state(environment, res_id_1, "skipped")
+
+    assert await check_resource_state(environment, res_id_2, "unavailable")
+
+    assert await check_resource_state(environment, res_id_3, "deployed")
+
+    log_doesnt_contain(
+        caplog, "inmanta.util", logging.ERROR, "An exception occurred while handling a future: 1 validation error for Event"
+    )

